### PR TITLE
OSASINFRA-3090: External LB support for on-prem platforms

### PR DESCRIPTION
**Pull request by EmilienM**
Support for External LB for BM, vsphere, OpenStack, ovirt and Nutanix platforms

### CODE REVIEW

**Comment by EmilienM** [comment]
/uncc AnnaZivkovic bostrt
/cc rvanderp3 jcpowermac  patrickdillon 

This is highly VIP now but I'll use it as a baseline and apply changes from the enhancement in this PR.

File: pkg/asset/ignition/machine/node.go
Comment (mandre) [inline pkg/asset/ignition/machine/node.go:38]: This conditional seems correct, we should probably add the same check to other on-prem platforms, in case we allow for empty APIVIPs.

File: pkg/tfvars/openstack/openstack.go
Comment (mandre) [inline pkg/tfvars/openstack/openstack.go:228]: Oh, I see. You have a typo, `openstack_external_loadbalancer` vs `openstack_external_load_balancer`.

**Review by mandre — COMMENTED** [review]
Seems like a good start. We'll need to see what type of validation we want to add.

I suggest we iterate on the patch with the OpenStack code first until we're happy with the result then port the same logic to other platforms.

**Comment by jcpowermac** [comment]
@EmilienM if loadBalancerType is external we can check for the existance of the LB and warn if not found.
I would rather you not remove at least that check for vsphere.

I thought this was tech preview? If so it should be added here:
https://github.com/openshift/installer/blob/master/pkg/types/validation/installconfig.go#L985-L1027

A little concerned we are changing how vSphere UPI is installed, @patrickdillon @rvanderp3 wdyt?
The positive for the change is all the onprem platforms will be the same, negative we are adding fields and CoreDNS that isn't required for vSphere UPI.

**Comment by EmilienM** [comment]
> @EmilienM if loadBalancerType is external we can check for the existance of the LB and warn if not found. I would rather you not remove at least that check for vsphere.

This was a mistake, I removed the revert.

> I thought this was tech preview? If so it should be added here: https://github.com/openshift/installer/blob/master/pkg/types/validation/installconfig.go#L985-L1027

Ok, I didn't know that interface. I'll add it here of course.

> A little concerned we are changing how vSphere UPI is installed, @patrickdillon @rvanderp3 wdyt? The positive for the change is all the onprem platforms will be the same, negative we are adding fields and CoreDNS that isn't required for vSphere UPI.

I agree it'll require your users to add `apiVIPs` and `ingresVIPs` fields. And that CoreDNS will be deployed, but I thought it was something we agreed in the enhancement...
I'll defer to @patrickdillon and his team but I think we should make the implementation consistent across all on-prem platforms...

**Comment by rvanderp3** [comment]
> > Right...but as currently implemented you are requiring this change in 4.13 even though this is tech preview.
> 
> I understand the concern and I don't know how to handle it, I'll defer to installer folks to make a decision. I'm ok with whatever option.

I think it makes sense to keep this behind the tech preview feature gate for 4.13.  @jcpowermac and I were under the impression this wasn't targeting 4.13 for GA.

**Comment by EmilienM** [comment]
> > > Right...but as currently implemented you are requiring this change in 4.13 even though this is tech preview.
> > 
> > 
> > I understand the concern and I don't know how to handle it, I'll defer to installer folks to make a decision. I'm ok with whatever option.
> 
> I think it makes sense to keep this behind the tech preview feature gate for 4.13. @jcpowermac and I were under the impression this wasn't targeting 4.13 for GA.

I updated my patch to reflect your proposal, so it's 100% backward compatible with the old behaviour.

**Comment by EmilienM** [comment]
There is still the `sets.String` error (see lint job) to fix , but I'll keep that for later, unless it's done by someone else. It's needed because apimachinery was updated in this PR.

File: pkg/types/vsphere/platform.go
Comment (EmilienM) [inline pkg/types/vsphere/platform.go:114]: should this be using a pointer for `VSpherePlatformLoadBalancer` ?

File: pkg/types/validation/installconfig.go
Comment (EmilienM) [inline pkg/types/validation/installconfig.go:1026]: I wanted to do:

```
if c.VSphere.LoadBalancer != nil {
  ...
}
```

But this doesn't work, I probably missed something obvious or is it because I'm not using a pointer?

File: pkg/types/vsphere/platform.go
Comment (patrickdillon) [reply pkg/types/vsphere/platform.go:114]: Yes, a pointer is preferred for an optional field.

File: pkg/types/validation/installconfig.go
Comment (patrickdillon) [reply pkg/types/validation/installconfig.go:1026]: > because I'm not using a pointer

+1

Comment (EmilienM) [reply pkg/types/validation/installconfig.go:1026]: thanks

File: pkg/types/vsphere/platform.go
Comment (EmilienM) [reply pkg/types/vsphere/platform.go:114]: thanks

**Comment by EmilienM** [comment]
Note for myself, add doc

File: pkg/asset/manifests/infrastructure.go
Comment (mandre) [inline pkg/asset/manifests/infrastructure.go:222]: We should set a default value for `installConfig.Config.VSphere.LoadBalancer.Type` to  `OpenShiftManagedDefault` in [SetPlatformDefaults()](https://github.com/openshift/installer/blob/master/pkg/types/vsphere/defaults/platform.go#L9) (or [installconfig.go](https://github.com/openshift/installer/blob/master/pkg/types/defaults/installconfig.go) if you want to do it globally for all on-prem platforms) so that this value is never empty, right? Or is there a reason to check for empty values?

File: pkg/tfvars/openstack/openstack.go
Comment (mandre) [inline pkg/tfvars/openstack/openstack.go:109]: How is empty a valid value? We should have validated install-config at this point, and set a valid default too.

That way, you only need to check for `installConfig.Config.Platform.OpenStack.LoadBalancer.Type == configv1.LoadBalancerTypeUserManaged` to get `userManagedLoadBalancer`.

File: pkg/asset/manifests/infrastructure.go
Comment (mandre) [inline pkg/asset/manifests/infrastructure.go:157]: Shouldn't the installer always fill the load balancer config into the PlatformStatus? We should set a default for installConfig.Config.BareMetal.LoadBalancer so that we it's always set and remove the above test.

Comment (EmilienM) [reply pkg/asset/manifests/infrastructure.go:157]: Since the feature is in TechPreview and it's not in the stable CRD, we have a validation later that checks for the install-config field to be set only if we are deploying with TechPreview featureSet. Therefore, I think we need this check for now and we should be able to remove it once it's promoted to stable.

File: pkg/asset/manifests/infrastructure.go
Comment (EmilienM) [reply pkg/asset/manifests/infrastructure.go:222]: Like I explained later, this field doesn't exist on stable CRD (see my API PR), so I don't think we can set a default here usable by both stable & TechPreview deployments. I think we should leave it that way for now.

Maybe we can confirm with @patrickdillon whether it's good practice or not.

File: pkg/tfvars/openstack/openstack.go
Comment (EmilienM) [reply pkg/tfvars/openstack/openstack.go:109]: Right. I'll remove the empty case. Good point.

File: pkg/asset/manifests/infrastructure.go
Comment (mandre) [reply pkg/asset/manifests/infrastructure.go:222]: Oh good point, it didn't occur to me that because it was tech preview we didn't want to set platform status for clusters unless they enabled the tech preview. That said, allowing empty values is confusing as hell - can we use a default _when_ the user enables tech preview?

Comment (EmilienM) [reply pkg/asset/manifests/infrastructure.go:223]: The default is at the API level, I'm not sure we want/need a default here.

File: data/data/install.openshift.io_installconfigs.yaml
Comment (rvanderp3) [inline data/data/install.openshift.io_installconfigs.yaml:3462]: Perhaps we want to mention in the description that this is tech preview

Comment (EmilienM) [reply data/data/install.openshift.io_installconfigs.yaml:3462]: can we? I literally copy/pasted from the original CRD.

**Review by pierreprinetti — COMMENTED** [review]
IIUC, this removes the API VIP and the FIPs when you set the external LB in install-config?

LGTM, with a couple nits on the testing part.

File: scripts/openstack/manifest-tests/user-managed-lb/test_cluster-infra.py
Comment (EmilienM) [reply scripts/openstack/manifest-tests/user-managed-lb/test_cluster-infra.py:21]: a copy/paste, I'll fix it.

**Comment by EmilienM** [comment]
> IIUC, this removes the API VIP and the FIPs when you set the external LB in install-config?

This addresses what is described in https://github.com/openshift/enhancements/blob/master/enhancements/network/external-lb-vips.md

> LGTM, with a couple nits on the testing part.

:pray:

File: scripts/openstack/manifest-tests/user-managed-lb/test_cluster-infra.py
Comment (pierreprinetti) [reply scripts/openstack/manifest-tests/user-managed-lb/test_cluster-infra.py:21]: also: your test must be called `test_*`

**Review by pierreprinetti — APPROVED** [review]

File: pkg/types/validation/installconfig.go
Comment (pierreprinetti) [inline pkg/types/validation/installconfig.go:1049]: What do you think of `field.Required` instead of `Invalid`? Invalid seems to be about a malformed value, while here it's empty and that's the problem.

Comment (EmilienM) [reply pkg/types/validation/installconfig.go:1049]: very good idea!

File: data/data/install.openshift.io_installconfigs.yaml
Comment (patrickdillon) [reply data/data/install.openshift.io_installconfigs.yaml:3462]: The description for `loadBalancer` includes that this is TechPreview, so I think we're good here :+1:

File: pkg/types/validation/installconfig.go
Comment (patrickdillon) [inline pkg/types/validation/installconfig.go:1062]: These validations should probably be moved to the platform-specific validations ([example](https://github.com/openshift/installer/blob/master/pkg/types/openstack/validation/platform.go) or at least they shouldn't be coupled with the featureset validations. 

There should be one validation to enforce that the field can only be set when tech preview is enabled, which is what you have in the following `else` block. The rest of the validations don't need to worry about tech preview, because they should only be run when a value is set.

**Review by patrickdillon — COMMENTED** [review]
This looks great. I did leave a lot of nits, which you are free to ignore. I also didn't go into depth on the openstack specific bits. I think there are a couple of changes needed, but I think they should be rather straightforward. Thanks again!

**Comment by patrickdillon** [comment]
> > > Right...but as currently implemented you are requiring this change in 4.13 even though this is tech preview.
> > 
> > 
> > I understand the concern and I don't know how to handle it, I'll defer to installer folks to make a decision. I'm ok with whatever option.
> 
> I think it makes sense to keep this behind the tech preview feature gate for 4.13. @jcpowermac and I were under the impression this wasn't targeting 4.13 for GA.

@jcpowermac @rvanderp3 @EmilienM  sorry I missed that you were waiting for me to weigh in here. Reading through this, it is unclear to me whether we're on the same page about vsphere UPI running coreDNS. Is there a resolution to this?

From an install config API point of view, we can't start requiring a field like `APIVIPs` that wasn't required before. Although we could conditionally require it when a new optional field is set. 

It seems like the default value for vSphere (at least upi, but there isn't really a distinction) must be user managed. And the validation only enforces APIVIPs when the openshift managed default is selected. So if we're ok with that default, it seems like we're fine or am I missing something?

File: pkg/types/validation/installconfig.go
Comment (EmilienM) [reply pkg/types/validation/installconfig.go:1062]: I've put them here because vsphere had some validations for the Zones work (that have been removed in the meantime).
I'll move them per platform then.

**Comment by patrickdillon** [comment]
/lgtm cancel 

I think @EmilienM is going to push some new changes, so this would get lost anyway.

/approve

File: pkg/asset/manifests/infrastructure.go
Comment (EmilienM) [reply pkg/asset/manifests/infrastructure.go:159]: That seems to work indeed, thanks!

File: pkg/asset/manifests/infrastructure.go
Comment (EmilienM) [reply pkg/asset/manifests/infrastructure.go:210]: That seems to work indeed, thanks!

File: pkg/tfvars/openstack/openstack.go
Comment (EmilienM) [reply pkg/tfvars/openstack/openstack.go:109]: the default being OpenShiftManaged, I'll set `userManagedLoadBalancer = false` for ""

File: pkg/tfvars/openstack/openstack.go
Comment (EmilienM) [reply pkg/tfvars/openstack/openstack.go:112]: I'll leave the code as is for this one:
* An extra validation wouldn't hurt I think
* We have code above that which calls `panic` already, so I followed the pattern.

File: pkg/tfvars/openstack/openstack.go
Comment (EmilienM) [reply pkg/tfvars/openstack/openstack.go:114]: nice!

File: pkg/tfvars/openstack/openstack.go
Comment (EmilienM) [reply pkg/tfvars/openstack/openstack.go:232]: :+1:

**Comment by EmilienM** [comment]
Note for reviewers, I'm still working on moving the validations into the platform directories. I did BM already but I'm not done yet.

**Comment by EmilienM** [comment]
> > > > > Right...but as currently implemented you are requiring this change in 4.13 even though this is tech preview.
> 
> > @jcpowermac @rvanderp3 @EmilienM sorry I missed that you were waiting for me to weigh in here. Reading through this, it is unclear to me whether we're on the same page about vsphere UPI running coreDNS. Is there a resolution to this?
> 
> We decided that for now we don't change anything with CoreDNS:
> 
>     * vsphere UPI will still live without CoreDNS because the API VIPs are empty in PlatformStatus.
> 
>     * When using a User Managed LB (on any on-prem), CoreDNS will be deployed, except for vsphere as explained in the first point.
> 
> 
> > From an install config API point of view, we can't start requiring a field like `APIVIPs` that wasn't required before. Although we could conditionally require it when a new optional field is set.
> 
> ack.
> 
> > It seems like the default value for vSphere (at least upi, but there isn't really a distinction) must be user managed. And the validation only enforces APIVIPs when the openshift managed default is selected. So if we're ok with that default, it seems like we're fine or am I missing something?
> 
> I'm fine with it. I'll apply that change now.

In fact, since we had no validations for the VIPs before, I won't add it here in this PR. This needs to be discussed and I think it's out of scope for this effort.

**Comment by EmilienM** [comment]
/test e2e-metal-ipi
/test e2e-vsphere-upi
/test e2e-nutanix-sdn 

The OpenStack failures are not related.
BM, VSphere and Nutanix seems flacky, but let's retest in case.

**Comment by EmilienM** [comment]
/hold cancel
I saw installs working on vsphere, BM and openstack, I think this is good to go.

**Comment by EmilienM** [comment]
> /retest-required

we have to wait for https://github.com/openshift/origin/pull/27736

**Comment by EmilienM** [comment]
/hold cancel
no need to hold, CI jobs are voting anyway.

**Pull request merged**
